### PR TITLE
Add support for X-Forwarded-Host proxy header

### PR DESCRIPTION
**Pull request by m-martin-78**
X-Forwarded-From and X-Forwarded-Proto are already used, so should X-Forwarded-Host.

### CODE REVIEW

**Comment by splitbrain** [comment]
How do we know we can trust this header?

File: inc/init.php [hunk 1]
```diff
@@ -495,7 +495,11 @@ function getBaseURL($abs = null)
 
     //split hostheader into host and port
     if (isset($_SERVER['HTTP_HOST'])) {
-        $parsed_host = parse_url('http://' . $_SERVER['HTTP_HOST']);
+    	$cur_host = $_SERVER['HTTP_HOST'];
+    	if(isset($_SERVER['HTTP_X_FORWARDED_HOST'])) {
+    		$cur_host = $_SERVER['HTTP_X_FORWARDED_HOST'];
+    	}
```
Comment (fiwswe) [inline inc/init.php:501]: More compact doing effectively the same thing without the unnecessary assign if $_SERVER['HTTP_X_FORWARDED_HOST'] is set:
`$cur_host = $_SERVER['HTTP_X_FORWARDED_HOST'] ?? $_SERVER['HTTP_HOST'];`

**Comment by fiwswe** [comment]
> How do we know we can trust this header?

Why wouldn't we trust it? What would an attacker gain by forging this header?

It is commonly used in proxy setups.

**Comment by splitbrain** [comment]
I'm not sure. We're trusting a HTTP header to create a redirect. It feels like this could lead to an open-redirect vulnerability or possibly even some kind of phishing risk. A quick google turned up this issue in rails: https://github.com/rails/rails/issues/29893

Another link points to https://portswigger.net/research/practical-web-cache-poisoning (which I haven't read in detail so I don't know if it applies here).

**Comment by m-martin-78** [comment]
> How do we know we can trust this header?

You already trust X-Forwarded-Proto a few lines below (without the same effect, but it still can cause a downgrade from HTTPS to HTTP).

We can be a bit more coherent with the app configuration by checking if `$_SERVER['REMOTE_ADDR']` matches [prox](https://www.dokuwiki.org/config:trustedproxy), but this should be done for all `X-Forwarded-*` headers that are set by reverse proxies.

The reverse proxy is in charge of rewriting these headers from the original requests if the client is not in its own trusted proxies list.

**Comment by m-martin-78** [comment]
I added the trustedproxy regex match to ensure `X-Forwarded-Host` won't be used if request is coming from a public IP address by default (considering the default value of the `trustedproxy` config property).

**Comment by splitbrain** [comment]
I like that approach :+1:

File: inc/init.php [hunk 2]
```diff
@@ -495,7 +495,11 @@ function getBaseURL($abs = null)
 
     //split hostheader into host and port
     if (isset($_SERVER['HTTP_HOST'])) {
-        $parsed_host = parse_url('http://' . $_SERVER['HTTP_HOST']);
+        $cur_host = $_SERVER['HTTP_HOST'];
+        if (isset($_SERVER['HTTP_X_FORWARDED_HOST']) && preg_match('/'.$conf['trustedproxy'].'/', $_SERVER['REMOTE_ADDR'])) {
```
Comment (fiwswe) [inline inc/init.php:499]: You also need to check for `empty($conf['trustedproxy'])`.

Comment (m-martin-78) [reply inc/init.php:499]: Well, if it is empty, that just means every remote address matches the expression. In that case, this is a deliberate setting from the administrator, right? Or do you consider an empty value indicates no proxy is trusted?

File: inc/init.php [hunk 3]
```diff
@@ -495,7 +495,11 @@ function getBaseURL($abs = null)
 
     //split hostheader into host and port
     if (isset($_SERVER['HTTP_HOST'])) {
-        $parsed_host = parse_url('http://' . $_SERVER['HTTP_HOST']);
+        $cur_host = $_SERVER['HTTP_HOST'];
```
[the remark this answers is not in the document]
Comment (m-martin-78) [reply inc/init.php:498]: Yessir!

**Comment by m-martin-78** [comment]
Also, I was looking into the `clientIP` function in `inc/common.php` to make it match the same behaviour, but I'm a bit confused by the logic in there. Usually, if `REMOTE_ADDR` is a trusted proxy, it should only return either the first IP in `X-Forwarded-For` list or the value of `X-Real-IP`. Is there any reason this function can return multiple client IP addresses?

**Comment by m-martin-78** [comment]
My bad, I forgot about updating the tests to match the new behaviour...

**Comment by m-martin-78** [comment]
Hi, could you help me a bit about updating the tests? How can I ensure a Dokuwiki configuration value? Thanks!

**Pull request merged**